### PR TITLE
Factor out UrlSelector from MultiServerRetryInterceptor

### PR DESCRIPTION
**Pull request by uschi2000**

### CODE REVIEW

**Comment by uschi2000** [comment]
FYI @divath

File: okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/UrlSelectorImpl.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        return new UrlSelectorImpl(ImmutableList.copyOf(canonicalUrls.build()));
+    }
+
+    @Override
+    public Optional<HttpUrl> redirectTo(HttpUrl current, String redirectBaseUrl) {
+        return redirectTo(current, HttpUrl.parse(redirectBaseUrl));
+    }
+
+    private Optional<HttpUrl> redirectTo(HttpUrl current, HttpUrl redirectBaseUrl) {
+        return baseUrlFor(redirectBaseUrl).flatMap(baseUrl -> {
```
Comment (markelliot) [inline okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/UrlSelectorImpl.java:57]: Can we simplify a bunch of this if we require all urls to have the same base path at construction time?

Comment (uschi2000) [reply okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/UrlSelectorImpl.java:57]: I think you still need to do the prefix matching, so it probably doesn't make a big difference?

Comment (markelliot) [reply okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/UrlSelectorImpl.java:57]: Is it that where we do all this work we no longer know the distinction between the application's requested path and the URL we're adding it to?

Comment (uschi2000) [reply okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/UrlSelectorImpl.java:57]: yeah.

Comment (uschi2000) [reply okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/UrlSelectorImpl.java:57]: we construct the okhttp interceptor with the discovery-consumed service URLs. we then intercept a feign/retrofit - generated URL that contains richer paths, parameters, query parameters, etc. redirecting here means replacing its base path by another one.

**Review by markelliot — APPROVED** [review]

**Pull request merged**